helmctlr/helm.go

```diff
@@ -122,23 +143,66 @@ func (c Controller) installOrUpdate(r *unstructured.Unstructured) error {
 	}
 
 	rlsName := c.releaseName(r)
-	if c.releaseExists(rlsName) {
-		_, err = c.Helm.UpdateRelease(
-			rlsName,
-			c.ChartDir,
-			helm.UpdateValueOverrides(cr),
-			helm.UpgradeWait(c.Wait),
-			helm.UpgradeTimeout(c.WaitTimeout))
+
+	tiller := c.tillerRendererForCR(r)
+
+	// load chart
+	chart, err := chartutil.LoadDir(c.ChartDir)
+	if err != nil {
 		return err
 	}
-	_, err = c.Helm.InstallRelease(
-		c.ChartDir,
-		c.Namespace,
-		helm.ReleaseName(rlsName),
-		helm.ValueOverrides(cr),
-		helm.InstallWait(c.Wait),
-		helm.InstallTimeout(c.WaitTimeout))
-	return err
+	release, err := c.storage.Last(rlsName)
+	if err != nil || release == nil {
+		//install release
+		installReq := &services.InstallReleaseRequest{
+			Namespace: r.GetNamespace(),
+			Name:      rlsName,
+			Chart:     chart,
+			Values:    &cpb.Config{Raw: string(cr)},
+			Wait:      c.Wait,
+			Timeout:   c.WaitTimeout,
+		}
+		//TODO: useful status from response?
+		_, err := tiller.InstallRelease(context.TODO(), installReq)
+		if err != nil {
+			return err
+		}
+	} else {
+		//update release
+		updateReq := &services.UpdateReleaseRequest{
+			Name:    rlsName,
+			Chart:   chart,
+			Values:  &cpb.Config{Raw: string(cr)},
+			Wait:    c.Wait,
+			Timeout: c.WaitTimeout,
+		}
+		//TODO: useful status from response?
+		_, err := tiller.UpdateRelease(context.TODO(), updateReq)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// tillerRendererForCR creates a ReleaseServer configured with a rendering engine that adds ownerrefs to rendered assets
+// based on the CR
+func (c Controller) tillerRendererForCR(r *unstructured.Unstructured) *tiller.ReleaseServer {
```

Comment:
Do you think we can test this method to be sure that it's correctly getting the ownerReferences of the passed in object and using those on the output objects? The `NewOwnerRefEngine` tests don't test the behavior of getting an unstructured object and getting it's ownerReferences. Possibly could just validate the created resources have it set correctly in the `TestResourceAddedHappyPathExists` test.

Reply:
Yes, I've validated this manually but should add a test for this

Reply:
Added!